docs/src/neural_ode/neural_gde.md

```diff
@@ -1,146 +1,259 @@
+This tutorial has been adapted from [here](https://github.com/CarloLucibello/GraphNeuralNetworks.jl/blob/master/examples/neural_ode_cora.jl).
 
-In this tutorial we will use Graph Differential Equations (GDEs) to perform classification on the [CORA Dataset](https://relational.fit.cvut.cz/dataset/CORA). We shall be using the Graph Neural Networks primitives from the package [GeometricFlux](https://github.com/yuehhua/GeometricFlux.jl).
+In this tutorial we will use Graph Differential Equations (GDEs) to perform classification on the [CORA Dataset](https://relational.fit.cvut.cz/dataset/CORA). We shall be using the Graph Neural Networks primitives from the package [GraphNeuralNetworks](https://github.com/CarloLucibello/GraphNeuralNetworks.jl).
 
-```julia
+```@example graphneuralode_cp
 # Load the packages
-using GeometricFlux, JLD2, SparseArrays, DiffEqFlux, DifferentialEquations
-using Flux: onehotbatch, onecold, throttle
-using Flux.Losses: logitcrossentropy
+using GraphNeuralNetworks, DifferentialEquations
+using DiffEqFlux: NeuralODE
+using GraphNeuralNetworks.GNNGraphs: normalized_adjacency
+using Lux, NNlib, Optimisers, Zygote, Random, ComponentArrays
+using Lux: AbstractExplicitLayer, glorot_normal, zeros32
+import Lux: initialparameters, initialstates
+using DiffEqSensitivity
 using Statistics: mean
-using LightGraphs: adjacency_matrix
+using MLDatasets: Cora
+using CUDA
+CUDA.allowscalar(false)
+device = CUDA.functional() ? gpu : cpu
 
 # Download the dataset
-download("https://rawcdn.githack.com/yuehhua/GeometricFlux.jl/a94ca7ce2ad01a12b23d68eb6cd991ee08569303/data/cora_features.jld2", "cora_features.jld2")
-download("https://rawcdn.githack.com/yuehhua/GeometricFlux.jl/a94ca7ce2ad01a12b23d68eb6cd991ee08569303/data/cora_graph.jld2", "cora_graph.jld2")
-download("https://rawcdn.githack.com/yuehhua/GeometricFlux.jl/a94ca7ce2ad01a12b23d68eb6cd991ee08569303/data/cora_labels.jld2", "cora_labels.jld2")
+dataset = Cora()
+
+# Preprocess the data and compute adjacency matrix
+classes = dataset.metadata["classes"]
+g = mldataset2gnngraph(dataset) |> device
+onehotbatch(data,labels)= device(labels).==reshape(data, 1,size(data)...)
+onecold(y) =  map(argmax,eachcol(y))
+X = g.ndata.features
+y = onehotbatch(g.ndata.targets, classes) # a dense matrix is not the optimal, but we don't want to use Flux here
+
+Ã = normalized_adjacency(g, add_self_loops=true) |> device
 
-# Load the dataset
-@load "./cora_features.jld2" features
-@load "./cora_labels.jld2" labels
-@load "./cora_graph.jld2" g
+(; train_mask, val_mask, test_mask) = g.ndata
+ytrain = y[:,train_mask]
 
 # Model and Data Configuration
-num_nodes = 2708
-num_features = 1433
-hidden = 16
-target_catg = 7
-epochs = 40
+nin = size(X, 1)
+nhidden = 16
+nout = length(classes)
+epochs = 20
+
+# Define the graph neural network
+struct ExplicitGCNConv{F1,F2,F3} <: AbstractExplicitLayer
+    Ã::AbstractMatrix  # nomalized_adjacency matrix
```

Comment:
I don't understand
```julia
julia> l = ExplicitGCNConv(Ã, nhidden => nhidden, relu)
ExplicitGCNConv(16 => 16, relu)  # 272 parameters

julia> l.Ã
2708×2708 CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}:
 0.25  0.0       0.0       0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.0  0.0       
 0.0   0.25      0.204124  0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.0  0.0       
 0.0   0.204124  0.166667  0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.0  0.0       
 0.0   0.0       0.0       0.5  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.0  0.0       
 0.0   0.0       0.0       0.0  0.166667  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.0  0.0       
 0.0   0.0       0.0       0.0  0.0       0.25  0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.0  0.0       
 0.0   0.0       0.0       0.0  0.0       0.0   0.2  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.0  0.0
 ⋮                                        ⋮                         ⋮              ⋱            ⋮                        ⋮                                  ⋮
 0.0   0.0       0.0       0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.333333  0.0       0.0  0.0  0.0  0.0  0.0       
 0.0   0.0       0.0       0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.333333  0.0  0.0  0.0  0.0  0.0       
 0.0   0.0       0.0       0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.5  0.0  0.0  0.0  0.0       
 0.0   0.0       0.0       0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.5  0.0  0.0  0.0       
 0.0   0.0       0.0       0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.5  0.0  0.0       
 0.0   0.0       0.0       0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.2  0.2       
 0.0   0.0       0.0       0.0  0.0       0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0       0.0       0.0  0.0  0.0  0.2  0.2
```
Could you elaborate

Reply:
`Ã` is already on GPU

Reply:
Right but device transfer functions are not defined on layers so `l |> cpu` or `l |> gpu` wont work.

Reply:
If we need to call `l |> cpu`. The definition here only serves the example. Although I'm curious how you would put it in the state, I can't find any info on this.

Reply:
Also, I wanna confirm this with you that `gpu(l)` does nothing to `l` if `l` is defined as a pure function such as `Dense`?

Reply:
That's a Flux workflow. In Lux, neural networks are structures with no values, so `l |> cpu` doesn't make sense because there is nothing to move to the CPU.

Reply:
Would make more sense to put `A` in the state if `initialstates` accepts extra arguments so I can pass `A`. Or we need to manually put `A` there after `setup`.

Reply:
The way to do it would be pass in a `init_A` function `() -> A` to the layer and define `initialstates(....) = (A = layer.init_A(), )`.

Also, it seems there is a documentation issue with Lux. I will do a small migration guide from Flux to Lux to document how to do these changes.

Reply:
So `init_A` just returns the global variable `A`? Would it be the same as 

`initialstates(rng::AbstractRNG, layer::ExplicitGCNConv) = (A=A,)`

Or you mean

`initialstates(rng::AbstractRNG, layer::ExplicitGCNConv, init_A) = (A=init_A(layer),)`

Reply:
You are right. It took me a bit of time to process your proposal. Will do as you said.

Reply:
Much clearer after reading it